File: mobile/lib/models/contract.dart
enum ContractStatus {
  draft('draft'),
  active('active'),
  expired('expired'),
  terminated('terminated'),
  cancelled('cancelled');

  const ContractStatus(this.value);
  final String value;

  static ContractStatus fromString(String s) {
    return ContractStatus.values.firstWhere(
      (e) => e.value == s,
      orElse: () => ContractStatus.draft,
    );
  }

  String get label => switch (this) {
    ContractStatus.draft => 'Draft',
    ContractStatus.active => 'Active',
    ContractStatus.expired => 'Expired',
    ContractStatus.terminated => 'Terminated',
    ContractStatus.cancelled => 'Cancelled',
  };

  /// Lifecycle actions are only offered while the contract can still change
  /// state from the field. Terminal states (expired / terminated / cancelled)
  /// expose no actions.
  bool get isActionable =>
      this == ContractStatus.draft || this == ContractStatus.active;
}

enum ContractType {
  purchase('purchase'),
  service('service'),
  subscription('subscription'),
  lease('lease'),
  sla('sla'),
  msa('msa'),
  sow('sow'),
  other('other');

  const ContractType(this.value);
  final String value;

  static ContractType fromString(String s) {
    return ContractType.values.firstWhere(
      (e) => e.value == s,
      orElse: () => ContractType.other,
    );
  }

  String get label => switch (this) {
    ContractType.purchase => 'Purchase',
    ContractType.service => 'Service',
    ContractType.subscription => 'Subscription',
    ContractType.lease => 'Lease',
    ContractType.sla => 'SLA',
    ContractType.msa => 'MSA',
    ContractType.sow => 'SOW',
    ContractType.other => 'Other',
  };
}

class ContractLineItem {
  final String id;
  final int? lineNumber;
  final String? itemCode;
  final String? description;
  final double? quantity;
  final double? unitPrice;
  final double? total;
  final String? glAccount;

  ContractLineItem({
    required this.id,
    this.lineNumber,
    this.itemCode,
    this.description,
    this.quantity,
    this.unitPrice,
    this.total,
    this.glAccount,
  });

  factory ContractLineItem.fromJson(Map<String, dynamic> json) {
    return ContractLineItem(
      id: json['id'] as String,
      lineNumber: (json['line_number'] as num?)?.toInt(),
      itemCode: json['item_code'] as String?,
      description: json['description'] as String?,
      quantity: (json['quantity'] as num?)?.toDouble(),
      unitPrice: (json['unit_price'] as num?)?.toDouble(),
      total: (json['total'] as num?)?.toDouble(),
      glAccount: json['gl_account'] as String?,
    );
  }
}

/// Spend rolled up against a contract's spend limit. Null when the backend
/// doesn't compute one (e.g. on the list endpoint).
class ContractSpend {
  final double invoicedTotal;
  final int invoiceCount;
  final double? spendLimit;
  final double? remaining;
  final bool overLimit;

  ContractSpend({
    required this.invoicedTotal,
    required this.invoiceCount,
    this.spendLimit,
    this.remaining,
    required this.overLimit,
  });

  factory ContractSpend.fromJson(Map<String, dynamic> json) {
    return ContractSpend(
      invoicedTotal: (json['invoiced_total'] as num?)?.toDouble() ?? 0,
      invoiceCount: (json['invoice_count'] as num?)?.toInt() ?? 0,
      spendLimit: (json['spend_limit'] as num?)?.toDouble(),
      remaining: (json['remaining'] as num?)?.toDouble(),
      overLimit: json['over_limit'] as bool? ?? false,
    );
  }
}

class Contract {
  final String id;
  final String? contractNumber;
  final String? title;
  final String? description;
  final ContractType contractType;
  final ContractStatus status;
  final String? vendorId;
  final String? vendorName;
  final String? currency;
  final double? totalValue;
  final double? spendLimit;
  final bool notToExceed;
  final DateTime? startDate;
  final DateTime? endDate;
  final DateTime? signedDate;
  final bool autoRenew;
  final int? renewalTermMonths;
  final int? renewalNoticeDays;
  final String? paymentTerms;
  final List<ContractLineItem> lineItems;
  final ContractSpend? spend;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Contract({
    required this.id,
    this.contractNumber,
    this.title,
    this.description,
    required this.contractType,
    required this.status,
    this.vendorId,
    this.vendorName,
    this.currency,
    this.totalValue,
    this.spendLimit,
    this.notToExceed = false,
    this.startDate,
    this.endDate,
    this.signedDate,
    this.autoRenew = false,
    this.renewalTermMonths,
    this.renewalNoticeDays,
    this.paymentTerms,
    this.lineItems = const [],
    this.spend,
    required this.createdAt,
    this.updatedAt,
  });

  factory Contract.fromJson(Map<String, dynamic> json) {
    return Contract(
      id: json['id'] as String,
      contractNumber: json['contract_number'] as String?,
      title: json['title'] as String?,
      description: json['description'] as String?,
      contractType: ContractType.fromString(json['contract_type'] as String),
      status: ContractStatus.fromString(json['status'] as String),
      vendorId: json['vendor_id'] as String?,
      vendorName: json['vendor_name'] as String?,
      currency: json['currency'] as String? ?? 'USD',
      totalValue: (json['total_value'] as num?)?.toDouble(),
      spendLimit: (json['spend_limit'] as num?)?.toDouble(),
      notToExceed: json['not_to_exceed'] as bool? ?? false,
      startDate: json['start_date'] != null
          ? DateTime.parse(json['start_date'] as String)
          : null,
      endDate: json['end_date'] != null
          ? DateTime.parse(json['end_date'] as String)
          : null,
      signedDate: json['signed_date'] != null
          ? DateTime.parse(json['signed_date'] as String)
          : null,
      autoRenew: json['auto_renew'] as bool? ?? false,
      renewalTermMonths: (json['renewal_term_months'] as num?)?.toInt(),
      renewalNoticeDays: (json['renewal_notice_days'] as num?)?.toInt(),
      paymentTerms: json['payment_terms'] as String?,
      lineItems: (json['line_items'] as List?)
              ?.map((e) =>
                  ContractLineItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
      spend: json['spend'] != null
          ? ContractSpend.fromJson(json['spend'] as Map<String, dynamic>)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
    );
  }
}

File: frontend/src/lib/types/contract.ts
import type { MoneyAmount, MoneyString } from '$lib/utils/money';

// Types for the Contracts surface. Mirrors the JSON returned by the
// `/api/contracts` endpoints (backend `Contract.to_dict()`). Money fields
// arrive as numbers (or null); date fields are ISO date strings (or null).
import type { BadgeTone } from '$lib/components/ui/Badge.svelte';
import type { MessageKey } from '$lib/i18n/messages';

export type ContractType =
	| 'purchase'
	| 'service'
	| 'subscription'
	| 'lease'
	| 'sla'
	| 'msa'
	| 'sow'
	| 'other';

export const CONTRACT_TYPES: ContractType[] = [
	'purchase',
	'service',
	'subscription',
	'lease',
	'sla',
	'msa',
	'sow',
	'other'
];

// The contract type is a table cell AND a `<select>` option beside the
// translated status badge, so an English literal here reads as a gap in the
// row rather than a deliberate data value. Keyed like the status map below.
export const CONTRACT_TYPE_LABEL_KEYS: Record<ContractType, MessageKey> = {
	purchase: 'contracts.type.purchase',
	service: 'contracts.type.service',
	subscription: 'contracts.type.subscription',
	lease: 'contracts.type.lease',
	sla: 'contracts.type.sla',
	msa: 'contracts.type.msa',
	sow: 'contracts.type.sow',
	other: 'contracts.type.other'
};

/** The message key for a contract type, or `null` for an unrecognised one. */
export function contractTypeLabelKey(contractType: string): MessageKey | null {
	return CONTRACT_TYPE_LABEL_KEYS[contractType as ContractType] ?? null;
}

export type ContractStatus =
	| 'draft'
	| 'active'
	| 'expired'
	| 'terminated'
	| 'cancelled';

export const CONTRACT_STATUSES: ContractStatus[] = [
	'draft',
	'active',
	'expired',
	'terminated',
	'cancelled'
];

/**
 * The i18n key carrying each status label — never the English string itself.
 *
 * Both surfaces that render a contract status (the `/contracts` list page and
 * `ContractModal`) are inside the i18n extraction slice, so a hardcoded
 * English map here put a translated lifecycle button beside an untranslated
 * `Active` badge. `Record<ContractStatus, MessageKey>` makes a new status a
 * compile error rather than a blank badge, and `contract.test.ts` proves every
 * key exists in the catalogue.
 */
export const STATUS_LABEL_KEYS: Record<ContractStatus, MessageKey> = {
	draft: 'contracts.status.draft',
	active: 'contracts.status.active',
	expired: 'contracts.status.expired',
	terminated: 'contracts.status.terminated',
	cancelled: 'contracts.status.cancelled'
};

/**
 * The message key for a status, or `null` for one this frontend doesn't know
 * (the caller renders the raw value — visible and searchable — rather than a
 * blank badge). `Contract.status` is typed `string` because the API is the
 * source of truth.
 */
export function contractStatusLabelKey(status: string): MessageKey | null {
	return STATUS_LABEL_KEYS[status as ContractStatus] ?? null;
}

// Badge tone per status, so the list page and the modal can't tint the same
// status two different shades — which is exactly what they did (the modal's
// `.badge.active` and the list's differed by an alpha step).
export const STATUS_TONES: Record<ContractStatus, BadgeTone> = {
	draft: 'accent',
	active: 'success',
	expired: 'warning',
	terminated: 'danger',
	// Cancelled is the absence of a signal, not a weak one — flat, not tinted.
	cancelled: 'neutral'
};

export interface ContractLineItem {
	id: string;
	line_number: number | null;
	item_code: string | null;
	description: string | null;
	quantity: number | null;
	// `schemas/contract.py::ContractLineItemResponse` inherits these as bare
	// `Decimal | None`, and pydantic's DEFAULT JSON shape for a Decimal is an
	// exact STRING — unlike the `float`-typed header figures below. So they
	// arrive as `"1200.00"`, which the old `number` type quietly mis-declared.
	unit_price: MoneyString | null;
	total: MoneyString | null;
	gl_account: string | null;
}

export interface ContractSpend {
	invoiced_total: MoneyAmount;
	invoice_count: number;
	spend_limit: MoneyAmount;
	/**
	 * Headroom left under the limit, computed SERVER-side. Never re-derive it
	 * from `spend_limit - invoiced_total`; `over_limit` is the server's own
	 * verdict and is what the UI branches on.
	 */
	remaining: MoneyAmount;
	over_limit: boolean;
}

export interface Contract {
	id: string;
	contract_number: string;
	title: string | null;
	description: string | null;
	contract_type: ContractType;
	status: ContractStatus;
	vendor_id: string;
	vendor_name: string | null;
	currency: string;
	total_value: MoneyAmount;
	spend_limit: MoneyAmount;
	not_to_exceed: boolean;
	start_date: string | null;
	end_date: string | null;
	signed_date: string | null;
	auto_renew: boolean;
	renewal_term_months: number | null;
	renewal_notice_days: number;
	renewal_alert_sent_at: string | null;
	payment_terms: string | null;
	owner_user_id: string | null;
	file_url: string | null;
	file_key: string | null;
	terms: Record<string, unknown> | null;
	line_items: ContractLineItem[];
	spend: ContractSpend | null;
	created_at: string;
	updated_at: string;
}

// Writable line-item shape for ContractCreate / line-item edits. All fields
// optional — the backend fills line_number when omitted.
export interface ContractLineItemInput {
	line_number?: number;
	item_code?: string | null;
	description?: string | null;
	quantity?: number | null;
	/** Request side — the exact decimal text typed, never a JSON number. */
	unit_price?: MoneyString | null;
	total?: MoneyString | null;
	gl_account?: string | null;
}

// POST /api/contracts body. vendor_id + contract_number are required.
export interface ContractCreate {
	contract_number: string;
	vendor_id: string;
	title?: string | null;
	description?: string | null;
	contract_type?: ContractType;
	currency?: string;
	total_value?: MoneyString | null;
	spend_limit?: MoneyString | null;
	not_to_exceed?: boolean;
	start_date?: string | null;
	end_date?: string | null;
	signed_date?: string | null;
	auto_renew?: boolean;
	renewal_term_months?: number | null;
	renewal_notice_days?: number;
	payment_terms?: string | null;
	owner_user_id?: string | null;
	line_items?: ContractLineItemInput[];
}

export interface ContractListResponse {
	items: Contract[];
	total: number;
	page: number;
	page_size: number;
}
